### PR TITLE
Improve performance of DataObject magic calls and addData()

### LINKED ISSUE #37637: [Issue] Improve performance of DataObject magic calls and addData()
This issue is automatically created based on existing pull request: magento/magento2#36345: Improve performance of DataObject magic calls and addData()

---------
The magic calls from the DataObject are being called over a 100k when there is no block_cache available and depending on the page often thousands of times when all block_cache is available.

The small change for addData() is just included because its a tiny optimization and prevents a bug from happening. When a $value is an array the original call to setData would override the internal $_data array, this is not expected behavior and just setting it directly to $_data is faster anyway.

The other change is in the __call magic method and the _underscore function and the change is quite radical. Since this method can be called so often, we want to prevent further function calls if possible. Functional calls are normally not that expensive, but will become more expensive if you use this extensively (thousands of even hundred of thousands extra calls):

These are my own benchmarks based on a local environment with opcache and with PROD mode. 

HOMEPAGE transaction
All cache enabled except for FPC (__call is called 975 times): 4.318307ms -> 1.469747ms All cache enabled except for FPC + Block_html  (__call is called 100379 times): 385.36412ms -> 88.412914ms

PLP transaction:
All cache enabled except for FPC  (__call is called 4628 times): 19.625463ms -> 5.754514ms All cache enabled except for FPC + Block_html (__call is called 112591 times): 404.84066ms -> 99.390374ms

Now this PR also skips some checks. It assumes if a $method call start with 'g' it will be for 'get' for example, this is done in my PR to get the most optimal performance result but might not be preferred. I also removed an index check with $args[1] for the setter since it isn't used anywhere, but might also be something we would want to change.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by igorwulff**
The magic calls from the DataObject are being called over a 100k when there is no block_cache available and depending on the page often thousands of times when all block_cache is available.

The small change for addData() is just included because its a tiny optimization and prevents a bug from happening. When a $value is an array the original call to setData would override the internal $_data array, this is not expected behavior and just setting it directly to $_data is faster anyway.

The other change is in the __call magic method and the _underscore function and the change is quite radical. Since this method can be called so often, we want to prevent further function calls if possible. Functional calls are normally not that expensive, but will become more expensive if you use this extensively (thousands of even hundred of thousands extra calls):

These are my own benchmarks based on a local environment with opcache and with PROD mode. 

HOMEPAGE transaction
- All cache enabled except for FPC (__call is called 975 times): 4.318307ms -> 1.469747ms 
- All cache enabled except for FPC + Block_html  (__call is called 100379 times): 385.36412ms -> 88.412914ms

PLP transaction:
- All cache enabled except for FPC  (__call is called 4628 times): 19.625463ms -> 5.754514ms 
- All cache enabled except for FPC + Block_html (__call is called 112591 times): 404.84066ms -> 99.390374ms

Now this PR also skips some checks. It assumes if a $method call start with 'g' it will be for 'get' for example, this is done in my PR to get the most optimal performance result but might not be preferred. I also removed an index check with $args[1] for the setter since it isn't used anywhere, but might also be something we would want to change.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#37637: Improve performance of DataObject magic calls and addData()

### CODE REVIEW

File: lib/internal/Magento/Framework/DataObject.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/DataObject.php]: I don't think we can check the only first letter. 
Examples: `$dataObject->greatMethod('test')` earlier were throwing exception, but now will go into this case.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/DataObject.php
Comment (igorwulff) [reply lib/internal/Magento/Framework/DataObject.php]: I tested and tried varied things, and reverted it back for now... performance gains are a bit less, but still very good:
Homepage without block_cache is now: 158.512655
PLP without block_cache is now: 172.213267

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by igorwulff** [comment]
Hi @ihor-sviziev We've been analyzing and testing this specific patch and think a change is required for the magic 'get' call, since we are seeing some issues in various areas, what do you think?

```
case 'get':
    if (isset($args[0]) && $args[0] !== null) {
        return $this->getData(
            self::$_underscoreCache[$method] ?? $this->_underscore($method),
                $args[0]
        );
    }

    return $this->_data[
        self::$_underscoreCache[$method] ?? $this->_underscore($method)
    ] ?? null;
```

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by igorwulff** [comment]
@magento run all tests

**Comment by igorwulff** [comment]
@magento run Static Tests

**Comment by igorwulff** [comment]
@magento run Unit Tests

**Comment by igorwulff** [comment]
@magento run Unit Tests

**Comment by igorwulff** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/DataObject.php
Comment (hostep) [inline lib/internal/Magento/Framework/DataObject.php]: Can we put this `$method[0].($method[1] ?? '').($method[2] ?? '')` into a variable? Not sure if this will have an extra performance impact but it should at least make the code a bit more readable 😄

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/DataObject.php
Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/DataObject.php]: I added a commend to make it more obvious and not affect performance improvement

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @igorwulff,
Looks like your changes broke the integration tests. Could you please double-check that?

**Comment by igorwulff** [comment]
> Hi @igorwulff, Looks like your changes broke the integration tests. Could you please double-check that?

I will look into it. 

I also saw an issue with vendor/magento/module-paypal/Model/Payflow/Request.php where the DataObject is extended. The logic in the overridden __call method does some weird things with changing the key when a = or & char is seen in the $args variable:
```
        if (isset($args[0]) && (strstr($args[0], '=') || strstr($args[0], '&'))) {
            $key .= '[' . strlen($args[0]) . ']';
        }
```

I don't understand why this is used like this.

It should be fixed by just changing the line:
```
$key = $this->_underscore(substr($method, 3));
```

To
```
$key = $this->_underscore($method);
```

**Comment by ihor-sviziev** [comment]
Hi @igorwulff,
Any updates on this? :)

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by igorwulff** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests

**Comment by engcom-Hotel** [comment]
Fixed static test failures. Functional and WebApi tests failures seem flaky to me.


@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests, Functional Tests CE

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Lima** [comment]
@magento run Functional Tests B2B , Functional Tests EE, Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento create issue

**Comment by engcom-Echo** [comment]
@engcom-Hotel moving this to `Pending Review` code approval is still pending for it.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@igorwulff, could you please double-check the performance impact after all changes were done recently to fix test failures?

**Comment by igorwulff** [comment]
Hi @ihor-sviziev the changes were minimal and mostly esthetic, so this did not impact the performance improvements. So I think we are good to go to proceed.

**Comment by engcom-Hotel** [comment]
Hello @igorwulff,

Thanks for the response!

As per your comment, we can move further with this PR, hence moving this PR to `Pending Review`.

Thanks

**Comment by hostep** [comment]
> @engcom-Hotel engcom-Hotel requested a review from hostep 20 hours ago

Am I supposed to do anything here? Kind of annoying that once you leave one little remark, you're supposed to do a full review 😝

**Review by ihor-sviziev — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
Hello @igorwulff,

While testing this PR we have below observations. 

In order to test the performance improvement we have used the below profiler:

https://github.com/NoiseByNorthwest/php-spx

### HomePage: With FPC Disable

_**2.4-develop**_
2 4-develop_homepage

**_After PR Changes_**
PR_homepage

### HomePage: With FPC and block_html cache Disable

_**2.4-develop**_
2 4-develop_homepage_fpc_blockhtml


**_After PR Changes_**
PR_homepage_fpc_blockhtml

### PLP: With FPC Disable

**_2.4-develop_**
2 4-develop_PLP

_**After PR Changes**_
PR_PLP

### PLP: With FPC and block_html cache Disable

**_2.4-develop_**
2 4-develop_PLP_fpc_blockhtml

_**After PR Changes**_
PR_PLP_fpc_blockhtml


As per the above screenshot, we can see the major improvements in the HomePage rendering after PR changes but it looks similar in the case of PLP after and before PR changes. 

We request you to please look into it. Also please let us know the tool used by you for performance improvement.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @igorwulff,

We have tested this PR, and we can see performance improvements on the home page, but it seems there is not much performance improvement visible to us for PLP. 

Please go through the detailed comment here https://github.com/magento/magento2/pull/36345#issuecomment-1634060205.

Please let us know which tool you have used for performance measurements. Till the time we are moving this PR `On Hold`.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @igorwulff,

Thank you for your contribution!

As per the comment https://github.com/magento/magento2/pull/36345#issuecomment-1634060205, it seems that the PLP page needs more performance improvement. Hence we are closing this PR for now.

Please feel free to reopen or let us know when you are ready to work on it again, and we will be happy to re-open it.

Thank you!

**Comment by hostep** [comment]
So you are closing a proven solution just because it doesn't match exactly with the numbers in the description?

@ihor-sviziev what do you make of this? If it improves performance, even if not 100% exact the same as written in the description and causes no regressions, why don't we go through with it?

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Unit Tests, WebAPI Tests

**Comment by engcom-Hotel** [comment]
As per this comment https://github.com/magento/magento2/pull/36345#issuecomment-1634060205, there are some improvements visible in this PR for HomePage and PDP. 

Hence we can move forward with the PR, but due to some tests are still failing, moving this PR into `Extended Testing`.

Thanks

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE, Functional Tests EE, Unit Tests, WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests EE, WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests EE

**Comment by engcom-Hotel** [comment]
All tests are green, hence moving this PR into `Merge in Progress`

**Pull request merged**
